### ISSUE TITLE
treatment of irregular morphology

### ISSUE THREAD

**Issue by leoalenc** [delph-in/matrix#601]
Hi everybody! I'm a computational linguist with a background in grammar engineering within LFG-XLE and Finite-State Morphology (XFST, Foma). As a theoretical linguist, I also have familiarity with the HPSG formalism, and I have worked through the most part of Ann Copestake's book on the LKB system.  Presently, I'm implementing a [HPSG grammar for Portuguese](https://github.com/LR-POR/PorGram) with the Grammar Matrix in the context of a research project at EMAp/FGV in collaboration with @arademaker. In order to learn the system, I've implemented a toy grammar for English and Latin. 
Could anyone help me with a problem I detected implementing irregular forms like _slept_ by means of the irregs.tab file? I can't find the reason for the following error message:

> Error: 
> Rule PAST-PART-LEX-RULE_INFL_RULE returned from morph analyser is undefined (error in grammar loading?) caused by `the dog has slept'  

In my toy grammar, I have the following in the my-irregs.tab file:

```
"
; Irregs.tab is used to define irregular forms
; The syntax is as follows:
; irreg_form IRREG_TYPE reg_form
; e.g.:
; ate VERB_eat PAST

slept pst-lex-rule sleeped
slept past-part-lex-rule sleeped

"
```
In the file my-irules-02.tdl, I have refined the inflectional rules defined with the Grammar Matrix:

```
pst-suffix :=
;;%suffix (!e !eed) (e ed) (sleep slept)
%suffix (!e !eed) (e ed)
pst-lex-rule.

past-part-suffix :=
;;%suffix (!e !eed) (e ed) (sleep slept)
%suffix (!e !eed) (e ed)
past-part-lex-rule.
```
The above commented out code worked fine, but I would like to block generation of the regular forms _sleeped_, following Copestake p. 199-200. 
The grammar is in [this repository](https://github.com/LR-POR/tutorial/tree/main/grammars/english). The version of the grammar which caused the error was loaded with the script my-script-02.

**Comment by goodmami** [comment]
Hi, I don't recall all the details of irregulars, but looking at the [ERG](http://moin.delph-in.net/wiki/ErgTop) as an example I see the `reg_form` should be a lemma and not the regular morphology you want to block. E.g., *sleep* not *sleeped* below:

```
slept V_PST_OLR sleep
```

So for your grammar, I would try the following:

```
slept pst-lex-rule sleep
slept past-part-lex-rule sleep
```

And also I would take the `(sleep slept)` inflectional patterns out of your irules.

Another thing that's different with the ERG is that the irregulars specify an inflectional rule (`v_pst_olr`) and not the rule type. I don't know if this is important, but if the above isn't enough, maybe try changing the `pst-lex-rule` and `past-part-lex-rule` to `pst-suffix` and `pst-part-suffix`, respectively.

**Comment by leoalenc** [comment]
> Hi, I don't recall all the details of irregulars, but looking at the [ERG](http://moin.delph-in.net/wiki/ErgTop) as an example I see the `reg_form` should be a lemma and not the regular morphology you want to block. E.g., _sleep_ not _sleeped_ below:
> 
> ```
> slept V_PST_OLR sleep
> ```
> 
> So for your grammar, I would try the following:
> 
> ```
> slept pst-lex-rule sleep
> slept past-part-lex-rule sleep
> ```
> 
> And also I would take the `(sleep slept)` inflectional patterns out of your irules.
> 
> Another thing that's different with the ERG is that the irregulars specify an inflectional rule (`v_pst_olr`) and not the rule type. I don't know if this is important, but if the above isn't enough, maybe try changing the `pst-lex-rule` and `past-part-lex-rule` to `pst-suffix` and `pst-part-suffix`, respectively.

@goodmami, thanks for drawing my attention to how irregular morphology is encoded in the ERG. You are right, the _irregs.tab_ file map irregular forms to lemmas. I corrected this in my grammar and tried out your suggestion of changing `pst-lex-rule` and `past-part-lex-rule` to `pst-suffix` and `pst-part-suffix`, but this didn't work. I received an analogous error message:

> Error: 
> Rule PST-SUFFIX_INFL_RULE returned from morph analyser is undefined (error in grammar loading?) caused by `the dog fell'
 
I inspected the ERG code and tested the grammar on some examples. Everything worked fine. So I have two suggestions to the developers of the Grammar Matrix:

1) Check if a grammar generated by the Grammar Matrix supports the definition of the mappings in the _irregs.tab_ file. Such a grammar does contain an empty _irregs.tab_ with the following comments:
```

"
; Irregs.tab is used to define irregular forms
; The syntax is as follows:
; irreg_form IRREG_TYPE reg_form
; e.g.:
; ate VERB_eat PAST

"
```
Besides, the script for loading the grammar in the LKB contains the following commented out line:

`;(load-irregular-spellings (lkb-pathname (parent-directory) "irregs.tab"))`

2) Consider generating a more user friendly _irregs.tab_ template. This seems to be an easy task: for each orth-changing lexical rule in the _irules.tdl_ file, generated by the morpholy page of the questionnaire, a line according to the following template would be generated:

`irreg_form TYPE_OLR lemma`
  
Supporting specification of irregular morphology by means of a _irregs.tab_ file would be a great feature!

**Comment by goodmami** [comment]
The other thing is that the rules are getting `INFL_RULE` tacked onto the end, which causes the rule lookup to fail. I looked at your repository and I see you've created a `my-globals.lsp` file containing the following ([link](https://github.com/LR-POR/tutorial/blob/a9aaad12538203577929b4189d213b3315c2cd38/grammars/english/lkb/my-globals.lsp#L36-L38
)):

```lisp
(defparameter *lex-rule-suffix* "_INFL_RULE"
"creates the inflectional rule name from the information
in irregs.tab - for PAGE compatability")
```

That parameter is commented out in the standard `globals.lsp` file. I would try leaving it commented out and see what happens.

There's also a bit of instruction here: http://courses.washington.edu/ling567/2005/lab3.html (search for "irregs")

**Comment by leoalenc** [comment]
> The other thing is that the rules are getting `INFL_RULE` tacked onto the end, which causes the rule lookup to fail. 
> 
> ```lisp
> (defparameter *lex-rule-suffix* "_INFL_RULE"
> "creates the inflectional rule name from the information
> in irregs.tab - for PAGE compatability")
> ```
> 
> That parameter is commented out in the standard `globals.lsp` file. I would try leaving it commented out and see what happens.
> 


@goodmami, thanks for solving the puzzle! That parameter definition was the problem! For other beginners as myself implementing irregular morphology, it is noteworthy quoting the following passage from the [page](http://courses.washington.edu/ling567/2005/lab3.html) you suggested:

> If you're going to use irregs.tab, you'll need to edit the file lkb/globals.lsp to comment out the definition of *lex-rule-suffix*. Restart the LKB once you've done this.

I've corrected the grammar in the repo and [a lot of examples with irregular verb forms are now parsed](https://github.com/LR-POR/tutorial/blob/main/grammars/english/my-test_sentences-results.txt).

**Comment by olzama** [comment]
Do we consider this issue still open?

**Comment by leoalenc** [comment]
> Do we consider this issue still open?

@olzama, thank you for asking. I'll close the issue, since my question was fully answered, thanks to @goodmami. The problem with the template _irregs.tab_ file presently generated by the Grammar Matrix is addressed in #603.
